geth/node/peerpool.go

```diff
@@ -0,0 +1,247 @@
+)
+
+const (
+	foundTimeout    = 90 * time.Second
+	defaultFastSync = 3 * time.Second
+	defaultSlowSync = 1 * time.Minute
+)
+
+// Limits represent min and max amount of peers
+type Limits [2]int
+
+// NewPeerPool creates instance of PeerPool
+func NewPeerPool(config map[discv5.Topic]Limits, fastSync, slowSync time.Duration) *PeerPool {
+	return &PeerPool{
+		config:   config,
+		fastSync: fastSync,
+		slowSync: slowSync,
+	}
+}
+
+type peerInfo struct {
+	// discoveredTime last time when node was found by v5
+	discoveredTime mclock.AbsTime
+	// connected is true if node is added as a statis peer
+	connected bool
+	// dropped is true if node was dropped by p2p server
+	// it can become false only when kademlia will discover it again
+	dropped bool
+
+	node *discover.Node
+}
+
+// PeerPool manages discovered peers and connects them to p2p server
+type PeerPool struct {
+	// config can be set only once per pool life cycle
+	config   map[discv5.Topic]Limits
+	fastSync time.Duration
+	slowSync time.Duration
+
+	mu sync.RWMutex
+	// TODO split this into separate maps to avoid unnecessary locking
+	peers         map[discv5.Topic]map[discv5.NodeID]*peerInfo
+	syncPeriods   []chan time.Duration
+	subscriptions []event.Subscription
+	quit          chan struct{}
+
+	wg sync.WaitGroup
+}
+
+// init creates data structures used by peer pool
+// thread safety must be guaranteed by caller
+func (p *PeerPool) init() {
+	p.subscriptions = make([]event.Subscription, 0, len(p.config))
+	p.syncPeriods = make([]chan time.Duration, 0, len(p.config))
+	p.peers = make(map[discv5.Topic]map[discv5.NodeID]*peerInfo)
+	for topic := range p.config {
+		p.peers[topic] = make(map[discv5.NodeID]*peerInfo)
+	}
+}
+
+// Start creates discovery search query for each topic and consumes peers found in that topic
+// in separate loop.
+func (p *PeerPool) Start(server *p2p.Server) error {
+	if server.DiscV5 == nil {
+		return ErrDiscv5NotRunning
+	}
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	// 2 goroutines per each topic
+	p.wg.Add(len(p.config) * 2)
+	p.quit = make(chan struct{})
+	// sync periods are stored because we need to close them once pool is stopped
+	p.init()
+
+	for topic := range p.config {
+		topic := topic
+		period := make(chan time.Duration, 2)
+		p.syncPeriods = append(p.syncPeriods, period)
+		found := make(chan *discv5.Node, 10)
+		lookup := make(chan bool, 10)
+		events := make(chan *p2p.PeerEvent, 20)
+		subscription := server.SubscribeEvents(events)
+		p.subscriptions = append(p.subscriptions, subscription)
+		go func() {
+			server.DiscV5.SearchTopic(topic, period, found, lookup)
+			p.wg.Done()
+		}()
+
+		go func() {
+			p.handlePeersFromTopic(server, topic, period, found, lookup, events)
+			p.wg.Done()
+		}()
+	}
+	return nil
+}
+
+func (p *PeerPool) handlePeersFromTopic(server *p2p.Server, topic discv5.Topic, period chan time.Duration, found chan *discv5.Node, lookup chan bool, events chan *p2p.PeerEvent) {
```

Comment:
could add chan direction to signature `period <-chan time.Duration, found <-chan *discv5.Node, lookup <-chan bool, events <-chan *p2p.PeerEvent`